tonlabs-cli/src/call.rs

```diff
@@ -63,38 +65,219 @@ pub fn call_contract(
         wait_for_timeout: None,
         access_key: None,
     })
-    .map_err(|e| format!("failed to create tonclient: {}", e.to_string()))?;
-    
-    let abi = std::fs::read_to_string(abi)
-        .map_err(|e| format!("failed to read ABI file: {}", e.to_string()))?;
+    .map_err(|e| format!("failed to create tonclient: {}", e.to_string()))
+}
+
+fn prepare_message(
+    ton: &TonClient,
+    addr: &TonAddress,
+    abi: &str,
+    method: &str,
+    params: &str,
+    header: Option<String>,
+    keys: Option<String>,
+) -> Result<EncodedMessage, String> {    
     
     let keys = load_keypair(keys)?;
+
+    ton.contracts.create_run_message(
+        addr,
+        abi,
+        method,
+        header.map(|v| v.into()),
+        params.into(),
+        keys.as_ref(),
+        None,
+    )
+    .map_err(|e| format!("failed to create inbound message: {}", e))
+}
+
+fn print_encoded_message(msg: &EncodedMessage) {
+    println!();
+    println!("MessageId: {}", msg.message_id);
+    print!("Expire at: ");
+    if msg.expire.is_some() {
+        let expire_at = Local.timestamp(msg.expire.unwrap() as i64 , 0);
+        println!("{}", expire_at.to_rfc2822());
+    } else {
+        println!("unknown");
+    }
+}
+
+fn pack_message(msg: &EncodedMessage, method: &str) -> String {
+    let json_msg = json!({
+        "msg": {
+            "message_id": msg.message_id,
+            "message_body": hex::encode(&msg.message_body),
+            "expire": msg.expire
+        },
+        "method": method,
+    });
+
+    hex::encode(serde_json::to_string(&json_msg).unwrap())
+}
+
+fn unpack_message(str_msg: &str) -> Result<(EncodedMessage, String), String> {
+    let bytes = hex::decode(str_msg)
+        .map_err(|e| format!("couldn't unpack message: {}", e))?;
     
-    let ton_addr = TonAddress::from_str(addr)
-        .map_err(|e| format!("failed to parse address: {}", e.to_string()))?;
+        let str_msg = std::str::from_utf8(&bytes)
+        .map_err(|e| format!("message is corrupted: {}", e))?;
+
+    let json_msg: serde_json::Value = serde_json::from_str(str_msg)
+        .map_err(|e| format!("couldn't decode message: {}", e))?;
+
+    let method = json_msg["method"].as_str()
```

Comment:
Better use struct with `#[derive(Serialize, Deserialize)]`. Then encoding and decoding are reduced to 1 expression